Fix #3571: Create simulationDb API

e-carlin opened the pull request:
This API is for reading the simulation directory that contains
sirepo-data.json or lib directory but not runDir

This branch is based on issue/3564-statelessCompute-api. Review and merge #3565 before reviewing this.

robnagler commented on sirepo/job_api.py:
```diff
@@ -151,6 +151,29 @@ def api_runStatus():
+
+
+@api_perm.require_user
+def api_simulationDb():
```
I think we need a different name. This is about doing something that requires the simulation_db. It's similar to statelessCompute, but one level higher, so it should be a "compute" of some sort. statefulAnalysis, statefulCompute, statefulQuery, etc.

e-carlin replied:
l went with "statefulCompute". I think anaylsis should be left for things that need runDirs and the relationship between stateful and stateless is nice. daba63620

robnagler commented on sirepo/job_api.py:
```diff
@@ -151,6 +151,29 @@ def api_runStatus():
+        req_data=PKDict(
+            **sirepo.http_request.parse_post().req_data,
+        ).pkupdate(
+            computeJobHash='unused',
```
This is repeated.

e-carlin replied:
27fd4c7d2

robnagler commented on sirepo/pkcli/job_cmd.py:
```diff
@@ -238,6 +238,13 @@ def _do_sequential_result(msg, template):
 
+def _do_simulation_db(msg, template):
+    try:
+        return template_common.simulation_db_dispatch(msg.data)
```
This looks like a pattern to encapsulate. Perhaps these requests are actually one API with a method attached with attributes describing what level of "compute" they require.

e-carlin replied:
43797f40e . Thoughts? I'm shying away from using one API because I think that leads us back down the road of IOCTL.

robnagler commented on tests/template/elegant_test.py:
```diff
@@ -27,7 +27,7 @@ def test_file_iterator():
     from pykern.pkunit import pkeq
     data = _find_example('bunchComp - fourDipoleCSR')
-    v = LatticeUtil(data, _elegant()._SCHEMA).iterate_models(
+    v = LatticeUtil(data, _elegant().SCHEMA).iterate_models(
```
I don't think we should be exporting this constant for this purpose. A unit test can reach inside for _SCHEMA, but really it should ask sim_data.

e-carlin replied:
It is really exported for use [here](https://github.com/radiasoft/sirepo/pull/3595/files#diff-687323f8346dfdf343f1eee4795f4492f47f05a6fae712ee4b920bd63897e68aR579). That ok?

robnagler replied:
Do you mean [here](https://github.com/radiasoft/sirepo/pull/3595/files#diff-24496517cc6cc5369fc69a13a07158c876115ef9eec8ceb098a25b374b083f7fL748)? If so, this is not exported, it's part of an object. There's not particular reason to use the constant vs calling sim_data or simulation_db in a test. It's not a performance issue.

e-carlin replied:
Sorry - not sure why my link didn't have a better line number. It is exported for use inside of [`template.common.stateful_compute_dispatch`](https://github.com/radiasoft/sirepo/pull/3595/files#diff-687323f8346dfdf343f1eee4795f4492f47f05a6fae712ee4b920bd63897e68aR592)

robnagler commented:
Is this ready to merge? Sorry I got behind, and now there are merge conflicts.

e-carlin commented:
Np - I've fixed the conflicts and done some testing again. Ready.

robnagler commented on tests/template/elegant_test.py:
```diff
@@ -42,8 +42,10 @@ def _elegant():
         method='rpn_value',
+        simulationType='elegant',
         value=expr,
         variables=variables or {},
```
Change to PKDict() when you touch code like this.

e-carlin replied:
e3c31a2ed

robnagler commented:
I thought this was merged. Shoudl it be?

robnagler commented:
We should merge this, but there are conflicts. I think the concept is sound.

e-carlin commented:
Ok, will fix today and report back when done.

e-carlin commented:
@robnagler all merge conflicts fixed. Ready for another review.

This has been sitting for a while and has a > 0% chance of breaking things so I squashed into 1 commit and rebased on master. I think this makes the history easier to understand and makes it easier to rollback if needed. If you have the branch checked out locally you'll need to 
```
git checkout issue/3571-simulation-db-api
git pull # This will complain about problems from upstream. Next command will fix
git reset --hard origin/issue/3571-simulation-db-api # This will blow away all of your local changes (if there are any)
```

robnagler reviewed (approved):
This is going to be one of those that we just have to try out in dev after the merge.
